### PR TITLE
Support IGNORE NULLS for LEAD window function

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Closes #.
Related to #9055

## Rationale for this change
Expanding IGNORE NULLS support for LEAD window function

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/sqllogictest/test_files/window.slt [hunk 1]
```diff
@@ -4169,43 +4169,103 @@ select lag(a, 2, null) ignore nulls over (order by id desc) as x1,
...
 FROM null_cases
 ORDER BY c2
 LIMIT 5;
 ----
-78
-63
-3
-24
-14
+78 50
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/window.slt:4267]: This only test is failing....

Comment (comphead) [reply datafusion/sqllogictest/test_files/window.slt:4267]: @mustafasrepo setting batch size to extreme low value will affect the window range?

File: datafusion/sqllogictest/test_files/window.slt [hunk 2]
```diff
@@ -4169,36 +4169,96 @@ select lag(a, 2, null) ignore nulls over (order by id desc) as x1,
...
+63 38
+3 53
+24 31
+14 94
 
 # result should be same with above, when lag algorithm work with pruned data.
 # decreasing batch size, causes data to be produced in smaller chunks at the source.
 # Hence sliding window algorithm is used during calculations.
+# LEAD function to be added, there is some bug with incoming data array when LEAD is called
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/window.slt:4257]: I'll address this in following PR, what I found is the incoming data is incomplete for LEAD.
so for  `set datafusion.execution.batch_size = 8000;`
the incoming array is complete 
```
[datafusion/physical-expr/src/window/lead_lag.rs:280:17] &array = StringArray
[
  "x",
  null,
  "b",
  null,
]
```
and for `set datafusion.execution.batch_size = 1;`
the data is partially missed
```
[datafusion/physical-expr/src/window/lead_lag.rs:280:17] &array = StringArray
[
  "x",
  null,
  "b"
]
```
Probably the error is in `built_in.rs` I'll check it in following PR

**Review by mustafasrepo — APPROVED** [review]
Thanks @comphead for this PR. It is LGTM!. Regarding failing test, you mentioned. The reason for its failure is that BoundedWindowExecutor uses `get_range` method to calculate required buffer for the result calculation. For most of the window function that uses window frame, `get_range` is not used. Because required range can be determined from window frame. However, for some builtin window functions window frame is not used such as `LEAD, LAG, ROW_NUMBER`. For these ones `get_range` should be implemented to make sure that required buffer is always present for the result. I think bug is related to the `get_range` implementation for the `LEAD`. I will file a subsequent PR to solve this problem.

**Pull request merged**